### PR TITLE
[gemma3] `Gemma3ForConditionalGeneration` compatible with assisted generation

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Enables assisted generation on `Gemma3ForConditionalGeneration`. Most of the novelty consists of a more precise compile-compatible prefill detection. If the new pattern is desirable, I can push it into more models :)

(This also fixes flaky issues in `Gemma3Vision2TextModelTest::test_prompt_lookup_decoding_matches_greedy_search` seen on `main`)

### CODE REVIEW

File: src/transformers/configuration_utils.py [hunk 1]
```diff
@@ -1367,7 +1367,12 @@ def recursive_diff_dict(dict_a, dict_b, config_obj=None):
 )
 
 
-def layer_type_validation(layer_types: list[str]):
-    """Check that each entry in `layer_types` are allowed."""
+def layer_type_validation(layer_types: list[str], num_hidden_layers: Optional[int] = None):
```
Comment (gante) [inline src/transformers/configuration_utils.py:1372]: more precise check 🔍 (and BC)

File: src/transformers/generation/candidate_generator.py [hunk 2]
```diff
@@ -185,7 +185,7 @@ def __init__(
                 )
 
         # We need to roll back the cache in assisted generation, only DynamicCache is supported
-        self.generation_config.cache_implementation = None
+        self.generation_config.cache_implementation = "dynamic_full"
```
Comment (gante) [inline src/transformers/generation/candidate_generator.py:188]: the assistant model also can't have sliding window cache, as we need to roll back there as well

File: src/transformers/models/gemma3/modeling_gemma3.py [hunk 3]
```diff
@@ -927,11 +927,22 @@ def forward(
                 "past_key_values": past_key_values,
                 "position_ids": position_ids,
             }
-            if token_type_ids is not None and inputs_embeds.shape[1] != 1:
+            # NOTE: this `is_prefill` logic is not flawless, it fails when we're using a cache eagerly initialized
+            # (e.g. compiled prefill) AND `pixel_values` are not provided. Determining prefill in that case requires
+            # checking data values, which is not compile-compatible.
```
Comment (gante) [inline src/transformers/models/gemma3/modeling_gemma3.py:925]: Very important question: is this a limitation for any of our partners? (e.g. VLLM)

Note that the previous prefill detection (`inputs_embeds.shape[1]`) isn't perfect either, as it returns a false positive when we feed more than one token at once, resulting in a crash.

Essentially, we have to pick between:
1. failing to detect prefill when cache is eagerly initialized + no pixel values is provided
2. failing on compilation in speculative-decoding-like inputs (we can add `or ((inputs_embeds.shape[1] != 1 and is_torchdynamo_compiling()))` to the clause)

I think the former is less restrictive, as the user has to make two non-conventional changes that are uncommon in tandem (eager cache init is speed-oriented, manually provinding `inputs_embeds` is research-oriented) 🤗 

[If we like this pattern, we can also apply something similar in other places. For instance, to replace `cache_positions[0] == 0` checks and enable compilation]

File: tests/generation/test_utils.py [hunk 4]
```diff
@@ -673,10 +672,6 @@ def test_assisted_decoding_matches_greedy_search(self, assistant_type):
         # - assisted_decoding does not support `use_cache = False`
         # - assisted_decoding does not support `batch_size > 1`
 
-        # No idea why this cause problem!
-        if type(self).__name__ not in ["Gemma3nTextModelTest"]:
-            set_model_tester_for_less_flaky_test(self)
```
Comment (gante) [inline tests/generation/test_utils.py:680]: took me a LOT of time to figure out this was causing the test to fail 😭 

(This function changes `num_hidden_layers` to `1`, which may clash with `layer_types` when it exists. IMO, it introduces a very nasty side-effect.)

File: tests/generation/test_utils.py [hunk 5]
```diff
@@ -729,6 +726,7 @@ def test_assisted_decoding_matches_greedy_search(self, assistant_type):
                 "output_attentions": self.has_attentions,
                 "return_dict_in_generate": True,
                 "use_cache": True,
+                "cache_implementation": "dynamic_full",  # see d)
```
Comment (gante) [inline tests/generation/test_utils.py:731]: This forces the first `generate` call, a plain greedy decoding, to use the same cache type we'll use in assisted generation.

This should reduce the flakiness in this test -- prior to this change, one call may be using a sliding window attention, while the other uses a full attention

(the correct solution is to enable sliding window attention in speculative decoding, which is a todo)

File: tests/test_modeling_common.py [hunk 6]
```diff
@@ -228,8 +228,6 @@ def _test_eager_matches_sdpa_inference(
             ("cuda", True, torch.float16): 5e-3,
         }
 
-    set_model_tester_for_less_flaky_test(self)
```
Comment (gante) [inline tests/test_modeling_common.py:228]: (tests fail due to the stricter `layer_types` check)

If this test becomes flaky because of this change, then we should increase the tolerances 👀 This function simply reduces the number of layers, i.e. the depth of the model.

File: tests/generation/test_utils.py [hunk 7]
```diff
@@ -1181,7 +1185,6 @@ def test_generate_from_inputs_embeds(self, _, num_beams):
         """Tests that we can generate from `inputs_embeds` instead of `input_ids` in LLMs, VLMs, etc"""
         # When supported, tests that the decoder model can generate from `inputs_embeds` instead of `input_ids`
         # if fails, you should probably update the `prepare_inputs_for_generation` function
-        set_model_tester_for_less_flaky_test(self)
```
Comment (zucchini-nlp) [inline tests/generation/test_utils.py:1186]: just to make sure, is the deletion same as below and relates to `layer_types`?

File: src/transformers/generation/candidate_generator.py [hunk 8]
```diff
@@ -298,6 +298,10 @@ def _update_past_and_masks(
             )
             self.assistant_kwargs = _prepare_token_type_ids(self.assistant_kwargs, input_ids.shape[-1])
 
+            # This unsets `dynamic_full`, needed to initialize a new cache for the assistant. After the first forward
+            # pass on each generation, we reuse the cache instead.
+            self.generation_config.cache_implementation = None
+
```
Comment (zucchini-nlp) [inline src/transformers/generation/candidate_generator.py:304]: I see that we're changing several `self.generation_config` attributes at init time and not setting them back to defaults since it's a deepcopy. For my understanding, why is it needed to be set back for cache?

File: src/transformers/testing_utils.py [hunk 9]
```diff
@@ -1637,6 +1637,9 @@ def assert_screenout(out, what):
 
 
 def set_model_tester_for_less_flaky_test(test_case):
+    # NOTE: this function edits the config object, which may lead to hard-to-debug side-effects. Use with caution.
+    # Do not use in tests that rely on non-model objects initialized from the config, like KV caches.
```
Comment (vasqu) [inline src/transformers/testing_utils.py]: Would extend that note to anything that relies on hybrid patterns:
- Moe also sometime start later (e.g. from layer 4 and on)
- Linear attention hybrid, e.g. bamba, jamba, qwen3next
- Attention hybrids, e.g. gemma2, gemma3, llama4

File: tests/generation/test_utils.py (hunk 4 above)
Comment (vasqu) [reply tests/generation/test_utils.py:680]: Agree, especially also possibly on hybrid models (see my comment on the note extension)

File: src/transformers/models/gemma3/modeling_gemma3.py (hunk 3 above)
Comment (vasqu) [reply src/transformers/models/gemma3/modeling_gemma3.py:925]: I fear that this pattern might become cumbersome in the future and we will have to add edge cases condition by condition. Atp, it would be smarter imo to just introduce a separate attribute to all caches and set it properly - this is more effort probably but way more robust/stable imo.

BUT, just a suggestion and this is a delicate problems so needs some more eyes that are more familiar here

**Review by vasqu — COMMENTED** [review]
Just a few comments on my side, more of a set of suggestions

I let the others be the proper judges here :p

File: src/transformers/testing_utils.py (hunk 9 above)
Comment (vasqu) [reply src/transformers/testing_utils.py]: Also rope x nope and other patterns like these

File: src/transformers/models/gemma3/modeling_gemma3.py (hunk 3 above)
Comment (zucchini-nlp) [reply src/transformers/models/gemma3/modeling_gemma3.py:925]: We also have a huge and ugly condition to check prefill in Qwen-VL models, so I'm 100% pro of adding a tiny utility which handles most edge cases and can be called in one-line :)

Regarding gemma3: vLLM does not use our cache, instead the caching mechanism is on their own side. Same for `torch.export` which decouples LM and vision so we never will use `ConditionalLM` block. We should be fine with them

~~The only edge case I can think of is when users want to get speed-boost but some of their inputs do not contain images. It is still possible that some examples are text-only though I can't say how often it happens "in-the-wild".~~

EDIT: I am dumb lol, if the input has no images we don't even need to execute the next lines of code. the current way looks fine to me

File: src/transformers/generation/candidate_generator.py (hunk 8 above)
Comment (gante) [reply src/transformers/generation/candidate_generator.py:304]: Because we validate that we don't use two potentially conflicting arguments, `cache_implementation` and `past_key_values`. If both exist, we throw an exception. In the 1st forward pass we set `cache_implementation` to the value required by assisted generation, in the other forward passes we want to use the growing cache.

We can allow both to co-exist if they match, but we don't have yet logic to handle it 🤗

File: src/transformers/models/gemma3/modeling_gemma3.py (hunk 3 above)
Comment (gante) [reply src/transformers/models/gemma3/modeling_gemma3.py:925]: > I fear that this pattern might become cumbersome in the future and we will have to add edge cases condition by condition. Atp, it would be smarter imo to just introduce a separate attribute to all caches and set it properly - this is more effort probably but way more robust/stable imo.

@vasqu I **think** (would have to check) that would also be an option. Essentially, use a similar mechanism as in the static sliding window cache: check how many tokens the cache has seen (`self.cumulative_length`). That should work on all cases, while being compile-compatible.

@Cyrilvallez thoughts on this?

____________________

> if the input has no images we don't even need to execute the next lines of code

@zucchini-nlp if the images are prepared outside the model and passed through `inputs_embeds`, wouldn't we need the following block?

File: tests/generation/test_utils.py (hunk 7 above)
Comment (gante) [reply tests/generation/test_utils.py:1186]: yes they are all related -- the KV cache is intialized incorrectly because `layer_types` has more layers than `num_hidden_layers`.

In some tests it's mostly harmless, but it makes debugging confusing ("why are there empty cache layers? is the model not being initialized properly?")

File: src/transformers/models/gemma3/modeling_gemma3.py (hunk 3 above)
Comment (zucchini-nlp) [reply src/transformers/models/gemma3/modeling_gemma3.py:925]: Yes, we do, though it is a super niche case and as you noted, I doubt users eager initialize cache with inputs embeds

Comment (Cyrilvallez) [reply src/transformers/models/gemma3/modeling_gemma3.py:925]: Sorry for being late on this! Yes, I think that simply checking `past_kv.get_seq_length() == 0` is enough and should be compile-compatible no?

Comment (Cyrilvallez) [reply src/transformers/models/gemma3/modeling_gemma3.py:925]: Humm, not for static cache layers though no, is that the issue? Otherwise, checking if it's None (or creating an equivalent `is_initialized` or something) method that does it is IMO a good option, as we never early_init in the library (only for executorch)

File: tests/generation/test_utils.py (hunk 4 above)
Comment (Cyrilvallez) [reply tests/generation/test_utils.py:680]: Yeah, I'm always battling with it. Dynamically changing the Tester itself is super bad, as no one WILL EVER EXPECT IT. I was planning to remove it hahaha

Comment (Cyrilvallez) [reply tests/generation/test_utils.py:680]: Especially as then all subsequent tests are not running correctly once it has been called

Comment (Cyrilvallez) [reply tests/generation/test_utils.py:680]: Cannot stress how bad of an idea it is in general haha

File: src/transformers/generation/utils.py [hunk 10]
```diff
@@ -2018,11 +2022,18 @@ def _prepare_cache_for_generation(
...
+        # initialization should happen inside the model at prefill time.
         else:
-            model_kwargs[cache_name] = (
-                DynamicCache(**dynamic_cache_kwargs)
-                if not requires_cross_attention_cache
-                else EncoderDecoderCache(DynamicCache(**dynamic_cache_kwargs), DynamicCache(**dynamic_cache_kwargs))
+            model_kwargs[cache_name] = DynamicCache(**dynamic_cache_kwargs)
+
+        # TODO (joao): this logic is incomplete, e.g. `offloaded` should apply to both caches. Refactor this function
+        # to correctly pass parameterization to both caches.
```
Comment (Cyrilvallez) [inline src/transformers/generation/utils.py:2040]: This can very easily be infered from the attributes set on the existing cache instance 😁

**Review by Cyrilvallez — APPROVED** [review]
Nice, LGTM! Very happy to see you removed `set_model_tester_for_less_flaky_test` on quite a few places, as I was planning to remove it entirely! (Will make my life simpler when I do it 🤗)
In general, I agree with the others that it may indeed be easier to have a small util on Cache to detect if it's fully empty or not, in a fully compile-compatible way if we can, instead of piling conditions!

**Comment by zucchini-nlp** [comment]
Can we merge it soon?  It is causing red CI 🙏🏻

File: src/transformers/models/gemma3/modeling_gemma3.py (hunk 3 above)
Comment (gante) [reply src/transformers/models/gemma3/modeling_gemma3.py:925]: @Cyrilvallez yeah, currently static cache's get_seq_len is data-dependent, so not good for compilation if used as a branch condition.

Adding an `is_initialized` 👍

Comment (zucchini-nlp) [reply src/transformers/models/gemma3/modeling_gemma3.py:925]: btw, just got an issue about gemma3 and prefill. We missed one more case which is when users want to cache shared system prompt and continue generation from there. 

On that note, I don't remember why we apply the bidirectional mask only in prefill. We can use token types whenever they are available, and if it is text-only the type ids will be all zeros
